### PR TITLE
Fix for #4899 through 'Embed' feature

### PR DESCRIPTION
**Pull request by klimov-paul**
Fix for https://github.com/yiisoft/yii2/issues/4899 through 'Embed' feature.
It allows creating nested models or list of models if necessary.
`yii\mongodb\embed\*` may be transfered into the core framework.

### CODE REVIEW

**Comment by klimov-paul** [comment]
@yiisoft/core-developers, what do you think?

**Comment by alexandernst** [comment]
I'm willing to test this. Can I apply this cleanly on top of 2.0.3 ?

**Comment by klimov-paul** [comment]
Yes, it should be fine

**Comment by alexandernst** [comment]
I'm testing this, but I'm not sure if I'm doing something wrong, or I'm just not using it as I'm supposed to, but:

I have a 3 levels nested object in my MongoDB that looks like this:

    _id: 'asdsdasdaaw',
    type: ['sss', 'xxx'],
    credentials: {
        email: 'asdas@asdas.com',
        password: 'qqqqqqq',
        salt: 'rr23tr23trwetwet'
    }

I'm trying this:

	var_dump($person->credentials);
	var_dump($person->credentials['password']);
	$person->credentials['password'] = 'iiiiiiiii';   // <------- doesn't work, fails with:

`Indirect modification of overloaded property app\models\Person::$credentials has no effect`

**Comment by klimov-paul** [comment]
Use:
```php
$person->credentials->password
```

**Comment by alexandernst** [comment]
Same error.

**Comment by alexandernst** [comment]
Just to be sure, `Person` in my code looks like this:

    class Person extends CActiveRecord {
        public static function collectionName() {
            return "person";
        }

        public function attributes() {
            return ['_id', 'credentials', .....];
        }

        ...
    }

Should I change the `attributes` function somehow?

Also

    var_dump($person->credentials);

says that `credentials`  is an array. Shouldn't that be some kind of object/class/data-holder of some sort? StdClass perhaps?

**Comment by qiangxue** [comment]
Is the embed thing specific to mongodb? From `ContainerTrait` and `ContainerInterface` it seems they are generic.

I also have some questions/suggestions about the naming. Will discuss about them later. But here's a name that bugs me a lot: "embed" is a verb, which makes the names like `getEmbed()` sounds very weird.

File: embed/ContainerInterface.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * @link http://www.yiiframework.com/
+ * @copyright Copyright (c) 2008 Yii Software LLC
+ * @license http://www.yiiframework.com/license/
+ */
+
+namespace yii\mongodb\embed;
+
+/**
+ * ContainerInterface allows converting any field or property, which is an associative array
+ * or list of associative arrays into object or list pf objects correspondently.
```
Comment (samdark) [inline embed/ContainerInterface.php]: pf → of

File: embed/ContainerInterface.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * Embed objects will use the copy of the source data, so modifying of source field will not affect
+ * instantiated embed objects and vice versa.
+ * In order to synchronize values between embed entities and container use [[synchronizeWithEmbed()]] method.
```
Comment (samdark) [inline embed/ContainerInterface.php]: I'd name it `refreshEmbed` to be consistent with AR.

File: embed/ContainerInterface.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    /**
+     * Returns list of values from embed objects named by source fields.
+     * @return array embed values.
+     */
+    public function getEmbedValues();
+
+    /**
+     * Fills up own fields by values fetched from embed objects.
+     */
+    public function synchronizeWithEmbed();
```
Comment (samdark) [inline embed/ContainerInterface.php]: `refreshEmbed`?

Comment (klimov-paul) [inline embed/ContainerInterface.php]: This is not that simple. Synchronization can be done both ways: from source to embed objects and from embed objects to source.
`refreshEmbed` sounds modre like refreshing embed values from source.

**Comment by klimov-paul** [comment]
@alexandernst, have you actually declared any embed objects?
Check the unit test from this PR to see a usage example.

**Comment by klimov-paul** [comment]
@qiangxue 

> Is the embed thing specific to mongodb? From ContainerTrait and ContainerInterface it seems they are generic.

This feature created to serve nested (embed) objects and list of objects, which MongoDB and ElasticSearch operates. They store each database entry in a JSON format, which means each particular field can be an object or array.
I have composed this solution into a trait, because it can be used not only for MongoDB but also for ElasticSearch and perhaps for some other NOSQL solutions.
I try to keep it genereric first of all because in order to be reused in different extensions it should be declared inside the core or at separated extension.

> I also have some questions/suggestions about the naming. Will discuss about them later. But here's a name that bugs me a lot: "embed" is a verb, which makes the names like getEmbed() sounds very weird.

Yes, naming is a pain here. It is hard to me to find proper name for the feature itself as well as for particular methods and classes. Any suggestions and ideas about it are wellcome.

File: embed/ContainerInterface.php (hunk 3 above)
Comment (samdark) [inline embed/ContainerInterface.php]: OK.

**Comment by samdark** [comment]
Embed → Embedded?

**Comment by klimov-paul** [comment]
Well it is hard to tell. 'Embedded' does not sounds better to me - it is more cerrect from lingual point of view, but produces unecessary extra symbols.
I have considered 'nested', 'subEntiry',  'nestedEntiry'. Finding fitting one is uneasy.

To find a proper name, we need to start from how the declaration of the nested entity should sounds.
At the moment it is 'embed*', like 
```
public function embedProfile()
```
When good solution for declaration found - the rest of the naming should be clear.
Perhaps `getEmbed()`, `setEmbed()` should be renamed to `getEmbedValue()`, `setEmbedValue()` or `getEmbedEntiry()`, `setEmbedEntiry()`.

Also what about namespace naming?

File: embed/ContainerInterface.php (hunk 3 above)
Comment (klimov-paul) [inline embed/ContainerInterface.php]: Perhaps it could be `refreshFromEmbed()`?

**Comment by alexandernst** [comment]
@klimov-paul I still can't find how to use this, can I get some help, please?
This is what  I have currently: http://pastebin.com/efMJTVTZ

**Comment by klimov-paul** [comment]
Start from using `stdClass`. Do not declare publlic fields. Watch for naming collisions.

Should be:
```php
class Person extends CActiveRecord implements IdentityInterface
{
        use ContainerTrait;
 
        public function embedCredentialsModel() {
                return $this->hasEmbed("credentials", "stdClass");
        }
 
        public static function collectionName()
        {
                return 'person';
        }
 
        public function attributes()
        {
                return ['_id', 'short_id', 'credentials'];
        }
}
```

$person = Person::findOne(['short_id' => '64fg8e']);
var_dump($person->credentialsModel);

File: embed/ContainerInterface.php (hunk 3 above)
Comment (qiangxue) [inline embed/ContainerInterface.php]: Perhaps simply `syncEmbed`, but need to find a good alternative to `embed` first.

**Comment by alexandernst** [comment]
@klimov-paul Ok, I followed your tips, removed the `Credentials` class, removed the public properties, renamed the `embed` method.

Now `var_dump $person->credentials` says that it's an array, and `var_dump $person->credentials->password` fails (because `credentials` in an array instead of an object).

I also tried `credentials['password']`, which seems to work (as expected), but as soon as I try to `credentials['password'] = "asdasdasd";`  it fails.

What am I doing wrong?

**Comment by klimov-paul** [comment]
As I said whach for naming collisions!
`$person->credentials` return original array!
`$person->credentialsModel` return object!

**Comment by alexandernst** [comment]
@klimov-paul Oh, yup, that was it. It's working now.

May I suggest something? Defining a method for each nested document seems kind of too much code to write (compared to no code at all, that is), maybe something like this would work better:

	public function embedAttributes()
	{
		return [
			"cretendialsModel" => ["credentials", "stdClass"],
			"personal_infoModel" => ["personal_info", "stdClass"],
		];
	}

PHP has quite good Reflections capabilities, so why not squeeze them? :)

Anyways, really good job @klimov-paul ! Big KUDOS!

**Comment by klimov-paul** [comment]
I have thought about using single method do declare embed entities. But I prefer using separated methods.
This allow achievement of 2 goals:
- it is consistent with normal relations definitions for ActiveRecord, which this feature is close to.
- it it better for lazy loading in case developer prefer to use `Object::className()` method, like:
```php
public function embedCredentialsModel() {
    return $this->hasEmbed("credentials", CredentialsModel::className()); // class autoload is triggered here
}
```

**Comment by klimov-paul** [comment]
Still, maybe I am wrong and definition should be adjusted.
We need to consider it as well as naming.

It whould be nice to here opinion of people involved in original issue disscussion about this all.
@docsolver, @cebe, @tadaszelvys, @Sammaye

**Comment by alexandernst** [comment]
@klimov-paul I have another suggestion.

Make is  possible to work with non-existent values without failing. Example:

    var_dump($person->creds->password); // ok
    var_dump($person->creds->foo_bar); // fails trying to access an undefined property, maybe create a key "foo_bar" and assign it "null" instead?

**Comment by klimov-paul** [comment]
Yes, this could be adjusted. I suppose there should be an option inside `Mapping`, which allows embed entiry creation from null value.
Still this is not a main concern right now: structure and naming are much more important.

**Comment by alexandernst** [comment]
I  think `nested` sounds good, as the naming represents what one would find inside, aka, a nested object/array of data. Also, it doesn't create unnecessarily large words:

    public function nestedProfileModel() { ... }

Also

    $this->syncNested();
    $this->(set/get)Nested();

**Comment by Sammaye** [comment]
Sorry for the delay, I will look it now

**Comment by Sammaye** [comment]
My first thoughts is that this is what I imagined.

I disagree with the use of one function to define all nested configurations. This used to be the case for things like joins and that and I prefer it in Yii2 style tbh. Plus many techs speak of nested sets as a form of join.

Also the functionality method allows for functions on the nested class to be called to augment, so I like this way/

**Comment by Sammaye** [comment]
How does this work with validation?

Especially if no class is found for the nested?

**Comment by klimov-paul** [comment]
What do you mean?

**Comment by Sammaye** [comment]
If you take your address example from  tests/data/ar/Customer.php how would I validate the values in that are correct instead of assuming or using assert on all of them?

**Comment by klimov-paul** [comment]
You will need to create a model class for 'address' embed, specifying validation rules for its attributes.
Then you can always call its validation manually:
```php
$customer->address->validate();
```
We may also create a special validator, which can be assigned to the owner model and validate embed one. Still you may do so on your own without much effots:
```
public function rules()
{
    return [
        ['address', function() {$this->addressModel->validate()}]
    ];
}
```

**Comment by Sammaye** [comment]
Hmm, I still think this would be good if you could validate very simple subdocuments without needing an entire model.

What about validating and saving array nesting, for example:

    [1,2,3,4,5]

?

File: embed/ContainerTrait.php [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+ * class User extends Model implements ContainerInterface
+ * {
+ *     use ContainerTrait;
+ *
+ *     public $profileData = [];
+ *     public $commentsData = [];
+ *
+ *     public function embedProfile()
+ *     {
+ *         return $this->hasEmbed('profileData', 'Profile');
```
Comment (alexandernst) [inline embed/ContainerTrait.php]: @klimov-paul Can you add an example for how `Profile` class should look like?

Comment (alexandernst) [inline embed/ContainerTrait.php]: @klimov-paul I mean... aside from the obvious

class Profile extends Model implements ContainerInterface {
	use ContainerTrait;

	public $firstName;
	public $lastName;
}

should there be anything else?

Comment (alexandernst) [inline embed/ContainerTrait.php]: Also, maybe make it clear that the class should be in the form of it's full namespace path:

    return $this->hasEmbed("profileData", "app\\models\\Profile");

**Comment by alexandernst** [comment]
@klimov-paul Not sure if this is an expected behaviour:

If I modify a subdocument in a controller and then `$model->save()` it, everything gets updated in the database (mongodb) as expected, but if I modify a subdocument in `beforeSave()`, the changes don't get saved unless I call `$this->synchronizeWithEmbed()`.

**Comment by qiangxue** [comment]
@klimov-paul I made some name changes in this gist: https://gist.github.com/qiangxue/688f09ce11b272cff777
Could you take a look?

I'm also thinking to replace `mapEmbedded()` and `mapEmbeddedList()` with `Map::to()` and `Map::toList()`. How do you think about this idea?

**Comment by klimov-paul** [comment]
I am fine with renaming 'embed' -> 'embeded' all over `ContainerInterface` and `ContainerTrait`.
I suppose namespace should be renamed accordingly.
What about declaration method itself, you suppose it is fine to keep it as `embed*`, or there should be another convension here? Al least It sounds more or less logical: method says 'embed something'.

I whould prefer `hasEmbeded()` over `mabEmbeded()` for declaration method, so it will be consistent with Active Record relation declaration methods `hasOne()` and `hasMany()`. Still it is acceptable compromise keeping in mind renaming `embededExists()` to `hasEmbeded()`. So I have no objection. Using `Map::to()` amd `Map::toList()` will require extra class to be imported via `use` statement, while syntax will not benefit much, so I suppose it is not needed.

`syncEmbedded()` is too generic. It shoulds like it will repopulate embebed object valies with the ones got from owner model. I whould keep `syncWithEmbedded()` to avoid misunderstanding. Currenty I have not added method, which synchronize embeded values to the onwer ones, because I suppose it is not needed. But in the future we may receive a feature request for it, so it should be a name reserved for this case at least.

Having `getAllEmbedded()` method does not make much sense. It is just a same as getting all Active Record relations. It is unlikely to be used.

**Comment by qiangxue** [comment]
> I suppose namespace should be renamed accordingly.

I think so.

> What about declaration method itself, you suppose it is fine to keep it as embed*, or there should be another convension here? Al least It sounds more or less logical: method says 'embed something'.

Yes, since "embed" is a verb and "embedXyz" makes perfect sense.

> I whould prefer hasEmbeded() over mabEmbeded() for declaration method...

Now I see why you chose this name.  Yes, I made the change mainly because it is a bit confusing to have both "hasEmbedded" and "embeddedExists()". They seem to have the same meaning. And "hasEmbedded" sounds like it should return a boolean (note that we also have "setEmbedded" and "getEmbedded", which makes the situation different from that with "hasOne").

> syncEmbedded() is too generic. It shoulds like it will repopulate embebed object values with the ones got from owner object. I whould keep syncWithEmbedded() to avoid misunderstanding.

I think "syncEmbedded" and "syncWithEmbedded" have the similar meaning: both indicate the sync process is bidirectional. If you want to indicate the direction is from embedded object to the owner property, then "syncFromEmbedded" should be used.

**Comment by klimov-paul** [comment]
OK, this sounds good.
What about the actual place for the 'embeded\*' classes? Should we place those inside the core under `yii\embeded`? Or perhaps the separated extension should be composed for this?

**Comment by qiangxue** [comment]
To be honest, I'm not quite sure. I'd like to hear other @yiisoft/core-developers' opinions.

The feature is certainly a generic one and could be put into the core. But I'm not certain if there is a better design. Putting it into the core means it will assume more responsibilities.

**Comment by klimov-paul** [comment]
Renaming and refactoring done.
Still waiting for opinion from @cebe and @samdark.

**Comment by cebe** [comment]
I am very busy this week, will put it on my list and try to review it on the weekend.

**Comment by skworden** [comment]
Does this allow for updating and deleting specific sub/embedded document(s)? Does this also allow for multiple of the same type of sub/embedded document? 

i.e. will we be able to select Yii2 from frameworks and delete or update it?

``` 
'_id':'123',
'frameworks'{
     {
        '_id':'1',
       'name':'Yii'
      },
      {
        '_id':'2',
       'name':'Yii2'
      }
}
```

**Comment by klimov-paul** [comment]
> Does this allow for updating and deleting specific sub/embedded document(s)?

Updating sure. Deleting - unlikely. You may set embedded entiry to be null, which after `refreshFromEmbedded` will apply `null` value to the source attribute, but the attribute itself will remain.

> Does this also allow for multiple of the same type of sub/embedded document?

I am unsure I understand this. Perhaps `mapEmbeddedList()` should work for you.

> .e. will we be able to select Yii2 from frameworks and delete or update it?

Only iterating over embedded related to 'frameworks'. You may unset any array element from it - changes will be applied.

**Comment by alexandernst** [comment]
@klimov-paul I'm testing your latest commits and I think something got broken.

I have this in my controller:

    $person = Person::findOne(['short_id' => '64fg8e']);
    $person->credentialsModel->password = 'test';
    var_dump($person->credentialsModel); // <--- I see "test" here
    return $this->render("myview", [
        "person" => $person // I see the old password here
    ]);

I thought I'd need to call `refreshFromEmbedded()`, so I did so right after setting the password, but then the entire `credentials` field is shown as  an empty array inn my view.

Can you confirm if this is a bug or I'm doing something wrong?

**Comment by klimov-paul** [comment]
Unit tests cover all functionality. They are passing successfully. So it should be something at your side.

**Comment by klimov-paul** [comment]
Added spcial validator for the embedded models.

**Comment by alexandernst** [comment]
@klimov-paul Ok, found where the problem is. If  `Credentials` extends from `CActiveRecord` (`class CActiveRecord extends ActiveRecord`) it fails, but if it extends from `Model` it works fine. Why is that?

**Comment by klimov-paul** [comment]
No idea, try to trace your program down to find the source of the problem.

**Comment by alexandernst** [comment]
@klimov-paul Ok, let's see where I can get.

One  more thing, can we make it somehow not necessary to call `refreshFromEmbedded()` every time a subdocument is modified? I can't really see a use-case where you'd want your model not to be synced with the subdocuments.

**Comment by klimov-paul** [comment]
It is not necessary alredy. It should be called only if you need the synchronization.
Usually developer will work either with pure array attributes or with embedded objects, but unlikely be doing both at the same time.
`yii\mongodb\ActiveRecord` performs synchronization before actually saving data only.

**Comment by alexandernst** [comment]
@klimov-paul What about https://github.com/yiisoft/yii2-mongodb/pull/13#issuecomment-92092283 ?

**Comment by klimov-paul** [comment]
Should be fixed already

**Comment by alexandernst** [comment]
@klimov-paul Indeed, it is fixed now. Thank you  :)

**Comment by skworden** [comment]
Is there anyway to use the mongodb's $pull to delete the subdocuments manually if this won't do it?

**Comment by klimov-paul** [comment]
You may run any MongoDB command if you want, but no such things will be performed automatically.

**Comment by alexandernst** [comment]
@klimov-paul Sorry for bugging you so much, but I have yet another problem.

I have a collection with categories and each document (category) looks like this:

    {
        //some fields...
        name: {
            "en-US": "Name of category",
            "es-ES": "Nombre de categoria",
            ..... more languages....
        }
    }

I guess you already know where am I going with this. The problem with that is that I can't define a `namesModel` class as this:

    class Names extends Model {
        //no properties as they should be added dynamically
    }

 because I can't add a new property to the class every time a new language is added to the database.

That's why I tried using reflexions:

    class Names extends Model {
        public function __construct() {
            foreach( Yii::$app->languagepicker->languages as $lang ) { //this returns an array of langs
                $this->{$lang} = "";
            }
        }
    }

But that fails with `Setting unknown property: app\models\CategoryName::en-US`:

Stacktrace:

    ../vendor/yiisoft/yii2/base/Component.php 197
    ../models/Category.php – yii\base\Component::__set('en-US', '') 15
    app\models\CategoryName::__construct()
    ../vendor/yiisoft/yii2/di/Container.php – ReflectionClass::newInstanceArgs([]) 374
    ../vendor/yiisoft/yii2/di/Container.php – yii\di\Container::build(.....) 151
    ../vendor/yiisoft/yii2/BaseYii.php – yii\di\Container::get(....) 344
    ../vendor/yiisoft/yii2-mongodb/embedded/Mapping.php – yii\BaseYii::createObject(....) 127
    ../vendor/yiisoft/yii2-mongodb/embedded/Mapping.php – yii\mongodb\embedded\Mapping::createValue(app\models\Category) 87
    ../vendor/yiisoft/yii2-mongodb/embedded/ContainerTrait.php – yii\mongodb\embedded\Mapping::getValue(app\models\Category) 153

My question is: How can I, dynamically, define properties in a class that extends from your patches?

**Comment by skworden** [comment]
@alexandernst this was the same problem i just ran into using this and was about to ask the same question.  Can this add more than one sub document in the same embedded attribute. If so, can someone please give an example.

**Comment by alexandernst** [comment]
@skworden Your question is not related to mine. I'm asking for properties of a (sub)document created in runtime, while you're asking for sub-sub-documents, which is possible. Just declare the embed*() method in your sub-document and you'll have a sub-sub-document.

**Comment by oduncan10** [comment]
I believe what he's asking is how to create multiple iterations of a subdocument on the same level. for example, within a profile, you may have 2 or more contacts, they would be assigned an Id and you would be able to add as many as you like and remove just one if you so choose.  In other words a document called PERSON would have contact 1, contact 2, contact 3, and on a later date I may add contact 4, etc, Is this possible within the scope of this Fix? Or is there a more elegant way to accomplish this goal. Thank you.

**Comment by alexandernst** [comment]
@oduncan10 @skworden I'm still not sure if I understand what you're asking for exactly. Are you looking for an array?

    class Contact {
        public $phone, $address;
    }

    class Person {
        public $contact = [];
    }

    $person = new Person();
    $person->contact[] = new Contact();
    $person->contact[0]->phone = 987654321;
    $person->contact[0]->address = "Street 123, Town";
    $person->contact[] = new Contact();
    $person->contact[1]->phone = 123456789;

**Comment by oduncan10** [comment]
Ideally each contact would be accessible as an object. The goal is to accomplish the handling of multiple iterations of the same subdocument as objects with very little modification of the core code. I.e. a person document with 5 contacts (5 subdocument objects accessible via person->contact->id), I need to be able remove the 2rd contact, update the first, and add two more.

**Comment by skworden** [comment]
@alexandernst 

say those two contacts are saved how do you add another one (basicly an update of the contact collection subdoument contact). 
How would I know what the next array Key is without looping through all of them before inserting a new contact to the subdocument

i.e. 
$person->contact[2]->phone=123456789;

**Comment by alexandernst** [comment]
@oduncan10 How would you access 5 different documents using the same key? You need to have a different key (or property).

    $person->contact->phone; <---- access first document
    $person->contact->phone; <------ still accessing first document

    $person->contact2->phone; <------- access second document, but we're using a different property (contact vs contact2)

Can you write a simple code example about how you are imagining accessing those 5 documents so I can understand you better?

@skworden The same way you'd add a value to a plain php array?

    $my_array = ["a"];
    $my_array[] = "b";  // <---- my_array now contains ["a", "b"]
    $my_array[] = "c"; // <---- my_array now contains ["a", "b", "c"]

Note that you can also use `count($my_array)` if you want to know how many items are there in the array. Or I'm completely not understanding you, or you're asking things completely unrelated to this patchset, but rather to basic knowledge of PHP.

**Comment by skworden** [comment]
@alexandernst I understand how to count array items etc.  I have this working but not using this patchset.

using 

```
$person->contact->phone;

$person->contact2->phone; 
```
i wouldn't know how many contacts the user is going to insert.  what am I missing? how would you do this dynamically with an unknown number of contacts?

**Comment by alexandernst** [comment]
@skworden Told you already! An array

**Comment by skworden** [comment]
It wasn't working that way.  I will keep looking and see what I'm doing wrong because you keep repeating the same thing over and over and it's not working.

**Comment by alexandernst** [comment]
@skworden Let me build a real example, maybe I'm overlooking some problem.

**Comment by skworden** [comment]
@alexandernst please! I'm having trouble updating/deleting with this.  I can create, update, delete outside of this it outside of this patch but it's driving me crazy trying to do all of this with it.

**Comment by alexandernst** [comment]
@skworden So, as I was saying, an array would work just fine. Just use `mapEmbeddedList` instead of `mapEmbedded` for specifying the mapping of the property :)

**Comment by alexandernst** [comment]
@klimov-paul I tried asking in SO ([here](http://stackoverflow.com/questions/29669658/declare-class-property-at-runtime-in-yii2)), but no luck. Any suggestions about https://github.com/yiisoft/yii2-mongodb/pull/13#issuecomment-93565659 ?

**Comment by klimov-paul** [comment]
Overwite `__set()` method:
```php
class MyModel extends Model
{
    private $dynamicFields = [];

    public function __set($name, $value)
    {
        $this->dynamicFields[$name] = $value;
    }

    public function __get($name)
    {
        if (array_key_exists($name, $this->dynamicFields) {
            return $this->dynamicFields[$name] = $value;
        } else {
            return parent::__get($name);
        }
    }
}
```

**Comment by alexandernst** [comment]
Yes, it works perfectly :)

I tweaked if a little bit thou:

	private $dynamicFields;

	public function __construct() {
		$this->dynamicFields = array_keys(Yii::$app->languagepicker->languages);
	}

	public function __set($name, $value) {
		if (in_array($name, $this->dynamicFields)) {
			$this->dynamicFields[$name] = $value;
		} else {
			parent::__set($name, $value);
		}
	}

	public function __get($name) {
		if (in_array($name, $this->dynamicFields)) {
			return $this->dynamicFields[$name];
		} else {
			return parent::__get($name);
		}
	}

Note that I'm using `in_array` because `dynamicFields` is not an associative array, but a plain array.

**Comment by klimov-paul** [comment]
`yii\mongodb\embedded\Mapping::unsetSource` added to reduce memory usage

**Comment by alexandernst** [comment]
@klimov-paul Do you think we could get this in for 2.0.4? (I mean, if other yii devs think the code is ok)

**Comment by klimov-paul** [comment]
I hope so. No BC break is involved, so there are no obstacles for that.

**Comment by alexandernst** [comment]
@klimov-paul Oh, I almost forgot about this one: https://github.com/yiisoft/yii2-mongodb/pull/13#issuecomment-91340671 Can we get it before merging for 2.0.4?

**Comment by cebe** [comment]
I have reviewed the code and read through most of the discussion here and in the original issue.
The approach seems fine for mongodb but the problem is generic and could be solved in the BaseActiveRecord in core framework so all dbms that support nested objects can benefit from it.
This however means that it will take longer to integrate as the scope gets wider and more general.
PostgreSQL for example has also the ability to store json data in a column.

I know about some issue with nested documents in elasticsearch which I would like to test on a proposal implementation before it gets merged and needs to support BC.

There are also two different alternatives as far as I see.

The proposed approach which defines a separate access to existing attributes, which has the problem of needing a sync on save to save the correct data when it has changed (`refreshFromEmbedded`).

An alternative approach would be to replace the original data with an object that holds the data and provides the access interface for it. This model could implement JsonSerialize and ensure data gets saved correctly for this attribute.

**Comment by klimov-paul** [comment]
I have abandonded the idea if replacing the attributes because of two reasons:

1) developer may want to access pure attribute value without extra object involved

2) replacing attributes will make particular ActiveRecord implementation ot be a mess. Inside the actual data saving methods it will be hard to get actual attribute values. This will also may riun 'dirty attributes' feature.

At present state this solution is completely optional. Its attachment does not affect anything and does not cost much effots.

> has the problem of needing a sync on save to save the correct data

You will need to this one way or another. There must be a step in the program, which will convert objects back into plain array so they can be saved in the storage. `refreshFromEmbedded` just does this.

**Comment by alexandernst** [comment]
@klimov-paul Does this patchset implements recursive adding attributes to a model via `setAttributes`?
I'm asking because I tried and only first-level attributes are assigned. Example:

    $person = Person::findOne([...]);
    $person->setAttributes([
        "name" => "Foo",  // <--- set ok
        "age" => 42,  // <--- set ok
        "creds" => [
            "pass" => "123" // <--- not set
        ]
    ], false);
    $person->save();

The database (mongodb) only gets the values of `name` and `age`, but the array `creds` gets lost.

I also tried assigning the attributes to the sub-model:

    $person->credsModel->setAttributes(["pass" => "123"], false);

But that also doesn't assign the values. I think it could be because of how I declare the attributes in `credsModel` (as you told me [here](https://github.com/yiisoft/yii2-mongodb/pull/13#issuecomment-93694724)).

Maybe I'm doing something wrong? If so, can you tell me how to do it properly? Thank you!

**Comment by klimov-paul** [comment]
It does not.

**Comment by alexandernst** [comment]
@klimov-paul How should I set an array with keys/values to a sub-document, then? Is there any way or should I just save the keys/values one by one? Also, do you plan adding this functionality?

**Comment by sokrat** [comment]
You can set array to the pure atribute, not to the sub-model.

**Comment by skworden** [comment]
@alexandernst to add multiple subdocs to a single attribute you do it as a new array.  I created a function that creates a new subdoc (multiple of the same type of subdoc) and i can use it with any of my models. However, this doesn't use any of the proposed code and my subdoc is just a separate model (for validation, looks like you guys fixed this with the proposed code).  You can make it a new function in the Active record / query 
````
Yii::$app->mongodb->getCollection('yourCollectionName')->update(['_id' => $id], ['$addToSet' => ['yourSubDocField'=> $yourArrayofData]]);
````
This assumes you have a mongoID as the pk your  field and you just pass the main collections _id as $id.

Also, I just get the post values like $yourArrayofData =Yii::$app->request->post('MySubDocModel');

This will add as many as you want without affecting the other subdocuments.

You can also easily update/delete just a single sub-document without looping through all of them as well using findone and setting your cursor properly.  If it's set right it will only return the single sub-document making it easy for you to update/view/delete without  having to loop though all of your subdocs.  I've been using the functions (more or less queries) for the last few days and they seem to be working as expected. However, I've only used them about 1-2k times so no real test.

**Comment by alexandernst** [comment]
I just want to keep the conversation alive until there isn't anything more to discuss, so when can merge this.

Unless I'm missing something (if so, please comment saying it), there are just 2 main points left, which I'm not sure if they will get implemented or they are a "wontfix" things:

- [ ] Set values to keys that are not declared (aka, create new keys dynamically) See [this comment](https://github.com/yiisoft/yii2-mongodb/pull/13#issuecomment-93494134)
- [ ] Implement a `setAttributes` method that will merge existing keys/values and passed keys/values. A temporal solution was commented [here](https://github.com/yiisoft/yii2-mongodb/pull/13#issuecomment-94941894), but that requires manual code (get the array, `array_merge` it, then set the merged array back to the pure attribute).

What do you think @klimov-paul ?

**Comment by klimov-paul** [comment]
> Set values to keys that are not declared (aka, create new keys dynamically)

It is up to you how to declare embedded classes. If you wish them to be dynamic it is your task to do so.
You can always use `stdClass` it allows to do so. For the Yii classes it is uncommon to have dynamic fields list, `yii\base\Object` was designed in the way preventing usage of not declared properties.

> Implement a setAttributes method that will merge existing keys/values and passed keys/values.

No, it will not be included in this solution. You need to populate embedded models manually just as you do for related models in Active Record.

**Comment by alexandernst** [comment]
Ok then, that said, I think this is ready to be merged, right? Or is there anything else that should be addressed before merging? @klimov-paul

**Comment by klimov-paul** [comment]
No.
Currently team has no agreement on this. @cebe believes issue can be solved in more staritforward way, which perhaps it can.
Sinse PR is lack of expert reviews and opinions It will remain opened for now.
I am unsure if this will appear at 2.0.4, cause we are running out of time.

**Comment by alexandernst** [comment]
Ah, I see. @cebe do you plan working on this? IMHO this is a really important feature to have because right now Yii2 claims to support MongoDB, ElasticSearch (and the NoSQL part of PostreSQL?), but the support is actually really primitive.

**Comment by cebe** [comment]
I do not think we should hurry with this and force it into 2.0.4 release. We should do more work on this and find a good solution that we can support for long time. Note that once it is merged, it needs to keep BC so it has to be a stable and good working solution.

**Comment by alexandernst** [comment]
@cebe I agree. I'm not trying to push such a feature in 2.0.4 (but I'd be glad if it happens). I'm just trying to keep the discussion alive so we can all agree on how exactly such a feature should work.

Imho, it would be nice to agree on where the code should live (yii-core if I understand correctly your thoughts on this) and what API/methods should it have (or how exactly should it work). Most of the times, that is the hardest part.

**Comment by cebe** [comment]
Yes this is the hardest part and it requires quite some time to be developed. Will try to find this time the next weeks but I can not promise anything.

**Comment by skworden** [comment]
As far as the manual handling of multiple (of the same type in the same collection) sub docs here  are some controller actions kinda like the default YII2 CRUD generated ones (as close as i could get to them). I don't know if it will help any for a more permanent elegant solution. Maybe @alexandernst you can use them.

this is to handle them independently from the parent model. You can also of course merge them with the parent model and validate multiple etc.
````
public function actionCreateSubDoc($id) {
        //id would be the id of the parent model not the subdoc
        $model = new SubDocModel();
        $id = ($id instanceof MongoId) ? $id : new \MongoId($id);
        if ($model->load(Yii::$app->request->post()) && $model->validate()) {
            Yii::$app->mongodb->getCollection('collectionName')->update(['_id' => $id], ['$addToSet' => [$subdoc => array_merge(['_id' => new \MongoId()], Yii::$app->request->post('SubDocModel'))]]);
                return $this->redirect(['view', 'id' => (string) $id]);
        } else {
            return $this->render('create', [ 'model' => $model]);
        }
    }

    public function actionUpdateSubDoc($id) {
        // the subdoc will always be at key [0] that is why we dont have to loop though
        //all of the subdocs like people say to
        $model = $this->findSubModel($id);
        $model = new SubDocModel($model['subDocField'][0]);
        if ($model->load(Yii::$app->request->post()) && $model->validate()) {
            Yii::$app->mongodb->getCollection('collectionName')->update(['_id' => $model[_id], 'subDocField._id' => $id], ['$set' => ['subDocField.$' => array_merge(['_id' => $id], Yii::$app->request->post('SubDocModel'))]]);
            return $this->render(['view', 'id' => (string) $model[_id]]);
        }
        return $this->render('update', [ 'model' => $model]);
    }

    public function actionDeleteSubDoc($id) {
        $id = ($id instanceof MongoId) ? $id : new \MongoId($id);
        Yii::$app->mongodb->getCollection('collectionName')->update([], ['$pull' => [$subdoc => ['_id' => $id]]]);
        return $this->redirect('index');
    }

    protected function findSubModel($id) {
        $id = ($id instanceof MongoId) ? $id : new \MongoId($id);
        if (($model = Yii::$app->mongodb->getCollection('collectionName')->findOne(['subDocField' . '._id' => $id], ['sudDocField._id.$' => 1])) !== null) {
            return $model;
        } else {
            throw new NotFoundHttpException('The requested page does not exist.');
        }
    }
````
Hopefully this will get some gears turning for someone and be able to make a more elegant reusable solution.

**Comment by alexandernst** [comment]
@cebe A month passed already, so I decided I'd ping you just to check if you had time to look at this.

**Comment by alexandernst** [comment]
@cebe Almost another month passed since my last ping. Did you get any time to have a look at this?

**Comment by samdark** [comment]
As you can see, there was an [important fix commit](https://github.com/klimov-paul/yii2-mongodb/commit/3fc5d67c83cb60f6b72998a6031d91de121092c0) during that time.

**Comment by alexandernst** [comment]
@samdark Yes, I was aware of the commit @klimov-paul pushed, but I'm not really sure how did that change the global state of this PR.

**Comment by klimov-paul** [comment]
Since this will not be decided soon, I have released this functionality as a separated extension:
https://github.com/yii2tech/embedded

You can use, if you want.
We may adopt this extension later, if the core team will agree.

**Comment by alexandernst** [comment]
@klimov-paul Should a new issue be opened so we can discuss how to implement this? Also, should it be created here on in yii2 repo as it affects other components too (PostreSQL, ElasticSearch, etc).

**Comment by klimov-paul** [comment]
The issue is still opened:
https://github.com/yiisoft/yii2/issues/4899

**Pull request closed without merging**